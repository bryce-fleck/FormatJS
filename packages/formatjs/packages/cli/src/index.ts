export {default as extract, ExtractCLIOptions} from './extract';
