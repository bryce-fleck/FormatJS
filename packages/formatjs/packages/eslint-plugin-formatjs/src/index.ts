import enforceDescription from './rules/enforce-description';
import noCamelCase from './rules/no-camel-case';
import noEmoji from './rules/no-emoji';
import noMultiplePlurals from './rules/no-multiple-plurals';
import noOffset from './rules/no-offset';
import blacklistElements from './rules/blacklist-elements';
const plugin = {
  rules: {
    'enforce-description': enforceDescription,
    'no-camel-case': noCamelCase,
    'no-emoji': noEmoji,
    'no-multiple-plurals': noMultiplePlurals,
    'no-offset': noOffset,
    'blacklist-elements': blacklistElements,
  },
};

export type Plugin = typeof plugin;

module.exports = plugin;
