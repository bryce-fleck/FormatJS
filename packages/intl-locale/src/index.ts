import {
  getInternalSlot,
  toObject,
  getOption,
  setInternalSlot,
  setMultiInternalSlots,
  objectIs,
  invariant,
} from '@formatjs/intl-utils';
import {
  parse as parseUnicodeLocaleId,
  canonicalizeUnicodeLanguageId,
  canonicalize,
} from './parser';
import {
  UnicodeLocaleId,
  UnicodeLanguageId,
  UnicodeExtension,
} from './unicode-locale-id-types';
import {emit, emitLanguageId} from './emitter';

export interface IntlLocaleOptions {
  language?: string;
  script?: string;
  region?: string;
  calendar?: string;
  collation?: string;
  hourCycle?: 'h11' | 'h12' | 'h23' | 'h34';
  caseFirst?: 'upper' | 'lower' | 'false';
  numberingSystem?: string;
  numeric?: boolean;
}

const RELEVANT_EXTENSION_KEYS = ['ca', 'co', 'hc', 'kf', 'kn', 'nu'] as const;
type RELEVANT_EXTENSION_KEY = typeof RELEVANT_EXTENSION_KEYS[number];
type ExtensionOpts = Record<RELEVANT_EXTENSION_KEY, string>;

interface IntlLocaleInternal extends IntlLocaleOptions {
  locale: string;
  initializedLocale: boolean;
  ast: UnicodeLocaleId;
}

const __INTERNAL_SLOT_MAP__ = new WeakMap<Locale, IntlLocaleInternal>();

const NUMBERING_SYSTEM_REGEX = /[a-z0-9]{3,8}(-[a-z0-9]{3,8})*/gi;

function applyOptionsToTag(
  tag: string,
  options: IntlLocaleOptions
): UnicodeLocaleId {
  const ast = parseUnicodeLocaleId(tag);
  const language = getOption(
    options,
    'language',
    'string',
    undefined,
    undefined
  );
  const script = getOption(options, 'script', 'string', undefined, undefined);
  const region = getOption(options, 'region', 'string', undefined, undefined);
  if (language !== undefined) {
    ast.lang.lang = language;
  }
  if (region !== undefined) {
    ast.lang.region = region;
  }
  if (script !== undefined) {
    ast.lang.script = script;
  }
  canonicalizeUnicodeLanguageId(ast.lang);
  return ast;
}

function applyUnicodeExtensionToTag(
  ast: UnicodeLocaleId,
  options: ExtensionOpts,
  relevantExtensionKeys: typeof RELEVANT_EXTENSION_KEYS
) {
  let unicodeExtension: UnicodeExtension | undefined;
  let keywords: UnicodeExtension['keywords'] = [];

  for (const ext of ast.extensions) {
    if (ext.type === 'u') {
      unicodeExtension = ext;
      if (Array.isArray(ext.keywords)) keywords = ext.keywords;
    }
  }

  const result = Object.create(null);

  for (const key of relevantExtensionKeys) {
    let value, entry;
    for (const keyword of keywords) {
      if (keyword[0] === key) {
        entry = keyword;
        value = entry[1];
      }
    }
    invariant(key in options, `${key} must be in options`);
    const optionsValue = options[key];
    if (optionsValue !== undefined) {
      invariant(
        typeof optionsValue === 'string',
        `Value for ${key} must be a string`
      );
      value = optionsValue;
      if (entry) {
        entry[1] = value;
      } else {
        keywords.push([key, value]);
      }
    }
    result[key] = value;
  }
  if (!unicodeExtension) {
    if (keywords.length) {
      ast.extensions.push({
        type: 'u',
        keywords,
        attributes: [],
      });
    }
  } else {
    unicodeExtension.keywords = keywords;
  }
  canonicalize(ast);
  return result;
}

function addLikelySubtags(unicodeLangId: UnicodeLanguageId): UnicodeLanguageId {
  const {lang, script, region} = unicodeLangId;
  const likelySubtags = Locale.LIKELY_SUBTAGS;
  const match =
    likelySubtags[
      emitLanguageId({lang, script, region, variants: []}) as 'aa'
    ] ||
    likelySubtags[emitLanguageId({lang, region, variants: []}) as 'aa'] ||
    likelySubtags[emitLanguageId({lang, script, variants: []}) as 'aa'] ||
    likelySubtags[lang as 'aa'] ||
    likelySubtags[emitLanguageId({lang: 'und', script, variants: []}) as 'aa'];
  if (!match) {
    throw new Error(`No match for addLikelySubtags`);
  }
  const parts = match.split('-');
  return {
    ...unicodeLangId,
    lang: lang === 'und' || !lang ? parts[0] : lang,
    script: script || parts[1],
    region: region || parts[2],
  };
}

function removeLikelySubtags(
  unicodeLangId: UnicodeLanguageId
): UnicodeLanguageId {
  const max = addLikelySubtags(unicodeLangId);
  const {variants} = max;
  const trials: UnicodeLanguageId[] = [
    {lang: max.lang, variants: []},
    {lang: max.lang, region: max.region, variants: []},
    {lang: max.lang, script: max.script, variants: []},
  ];
  for (const trial of trials) {
    if (isLanguageEqualWithoutVariants(max, addLikelySubtags(trial))) {
      return {
        ...trial,
        variants,
      };
    }
  }
  return max;
}

function isLanguageEqualWithoutVariants(
  l1: Omit<UnicodeLanguageId, 'variants'>,
  l2: Omit<UnicodeLanguageId, 'variants'>
): boolean {
  return (
    l1.lang === l2.lang && l1.region === l2.region && l1.script === l2.script
  );
}

export class Locale {
  constructor(tag: string | Locale, opts?: IntlLocaleOptions) {
    // test262/test/intl402/RelativeTimeFormat/constructor/constructor/newtarget-undefined.js
    // Cannot use `new.target` bc of IE11 & TS transpiles it to something else
    const newTarget =
      this && this instanceof Locale ? this.constructor : void 0;
    if (!newTarget) {
      throw new TypeError("Intl.Locale must be called with 'new'");
    }

    const {relevantExtensionKeys} = Locale;

    const internalSlotsList: Array<keyof IntlLocaleInternal> = [
      'initializedLocale',
      'locale',
      'calendar',
      'collation',
      'hourCycle',
      'numberingSystem',
    ];

    if (relevantExtensionKeys.indexOf('kf') > -1) {
      internalSlotsList.push('caseFirst');
    }

    if (relevantExtensionKeys.indexOf('kn') > -1) {
      internalSlotsList.push('numeric');
    }

    if (typeof tag !== 'string' && typeof tag !== 'object') {
      throw new TypeError('tag must be a string or object');
    }

    if (
      typeof tag === 'object' &&
      getInternalSlot(__INTERNAL_SLOT_MAP__, tag, 'initializedLocale')
    ) {
      tag = getInternalSlot(__INTERNAL_SLOT_MAP__, tag, 'locale');
    } else {
      tag = tag.toString() as string;
    }

    let options: IntlLocaleOptions;
    if (opts === undefined) {
      options = Object.create(null);
    } else {
      options = toObject(opts);
    }

    const ast = applyOptionsToTag(tag, options);
    const opt = Object.create(null);
    const calendar = getOption(
      options,
      'calendar',
      'string',
      undefined,
      undefined
    );
    if (calendar !== undefined) {
      if (!NUMBERING_SYSTEM_REGEX.test(calendar)) {
        throw new RangeError('invalid calendar');
      }
    }
    opt.ca = calendar;

    const collation = getOption(
      options,
      'collation',
      'string',
      undefined,
      undefined
    );
    if (collation !== undefined) {
      if (!NUMBERING_SYSTEM_REGEX.test(collation)) {
        throw new RangeError('invalid collation');
      }
    }
    opt.co = collation;
    const hc = getOption(
      options,
      'hourCycle',
      'string',
      ['h11', 'h12', 'h23', 'h34'],
      undefined
    );
    opt.hc = hc;
    const kf = getOption(
      options,
      'caseFirst',
      'string',
      ['upper', 'lower', 'false'],
      undefined
    );
    opt.kf = kf;
    let kn = getOption(options, 'numeric', 'boolean', undefined, undefined);
    if (kn !== undefined) {
      kn = !!String(kn);
    }
    opt.kn = kn;
    const numberingSystem = getOption(
      options,
      'numberingSystem',
      'string',
      undefined,
      undefined
    );
    if (numberingSystem !== undefined) {
      if (!NUMBERING_SYSTEM_REGEX.test(numberingSystem)) {
        throw new RangeError('Invalid numberingSystem');
      }
    }
    opt.nu = numberingSystem;
    const r = applyUnicodeExtensionToTag(ast, opt, relevantExtensionKeys);
    setMultiInternalSlots(__INTERNAL_SLOT_MAP__, this, {
      locale: r.locale,
      calendar: r.ca,
      collation: r.co,
      hourCycle: r.hc,
    });
    if (relevantExtensionKeys.indexOf('kf') > -1) {
      setInternalSlot(__INTERNAL_SLOT_MAP__, this, 'caseFirst', r.kf);
    }
    if (relevantExtensionKeys.indexOf('kn') > -1) {
      setInternalSlot(
        __INTERNAL_SLOT_MAP__,
        this,
        'numeric',
        objectIs(r.kn, 'true')
      );
    }
    setInternalSlot(__INTERNAL_SLOT_MAP__, this, 'numberingSystem', r.nu);
    setInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast', ast);
  }

  /**
   * https://www.unicode.org/reports/tr35/#Likely_Subtags
   */
  public maximize(): Locale {
    const ast = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    try {
      const maximizedLang = addLikelySubtags(ast.lang);
      return new Locale(
        emit({
          ...ast,
          lang: maximizedLang,
        })
      );
    } catch (e) {
      return new Locale(emit(ast));
    }
  }

  /**
   * https://www.unicode.org/reports/tr35/#Likely_Subtags
   */
  public minimize(): Locale {
    const ast = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    try {
      const minimalLang = removeLikelySubtags(ast.lang);
      return new Locale(
        emit({
          ...ast,
          lang: minimalLang,
        })
      );
    } catch (e) {
      return new Locale(emit(ast));
    }
  }

  public toString() {
    return emit(getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast'));
  }

  public get baseName() {
    const {lang} = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    return [lang.lang, lang.script, lang.region, ...(lang.variants || [])]
      .filter(Boolean)
      .join('-');
  }

  public get calendar() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'calendar');
  }

  public get collation() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'collation');
  }

  public get hourCycle() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'hourCycle');
  }

  public get caseFirst() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'caseFirst');
  }

  public get numeric() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'numeric');
  }
  public get numberingSystem() {
    return getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'numberingSystem');
  }
  /**
   * https://tc39.es/proposal-intl-locale/#sec-Intl.Locale.prototype.language
   */
  public get language() {
    // Spec implies re-parsing but let's not do that
    const {lang} = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    return lang.lang;
  }
  /**
   * https://tc39.es/proposal-intl-locale/#sec-Intl.Locale.prototype.script
   */
  public get script() {
    const {lang} = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    return lang.script;
  }
  /**
   * https://tc39.es/proposal-intl-locale/#sec-Intl.Locale.prototype.region
   */
  public get region() {
    const {lang} = getInternalSlot(__INTERNAL_SLOT_MAP__, this, 'ast');
    return lang.region;
  }

  static relevantExtensionKeys = RELEVANT_EXTENSION_KEYS;

  static LIKELY_SUBTAGS: Record<string, string> = {};

  public static _addLikelySubtagData(data: Record<string, string>): void {
    Locale.LIKELY_SUBTAGS = data;
  }
}

try {
  if (typeof Symbol !== 'undefined') {
    Object.defineProperty(Locale.prototype, Symbol.toStringTag, {
      value: 'Intl.Locale',
      writable: false,
      enumerable: false,
      configurable: true,
    });
  }

  Object.defineProperty(Locale.prototype.constructor, 'length', {
    value: 0,
    writable: false,
    enumerable: false,
    configurable: true,
  });
} catch (e) {
  // Meta fix so we're test262-compliant, not important
}
