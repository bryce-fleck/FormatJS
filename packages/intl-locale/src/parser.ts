import {
  UnicodeLocaleId,
  UnicodeLanguageId,
  UnicodeExtension,
  TransformedExtension,
  PuExtension,
  OtherExtension,
  KV,
  Extension,
} from './unicode-locale-id-types';

const ALPHA_2_3 = /^[a-z]{2,3}$/i;
const ALPHANUM_1_8 = /^[a-z0-9]{1,8}$/i;
const ALPHANUM_2_8 = /^[a-z0-9]{2,8}$/i;
const ALPHANUM_3_8 = /^[a-z0-9]{3,8}$/i;
const ALPHA_5_8 = /^[a-z]{5,8}$/i;
const KEY_REGEX = /^[a-z0-9][a-z]$/i;

const TYPE_REGEX = /^[a-z0-9]{3,8}$/i;
const ALPHA_4 = /^[a-z]{4}$/i;
// alphanum-[tTuUxX]
const OTHER_EXTENSION_TYPE = /^[0-9a-svwyz]$/i;
const UNICODE_REGION_SUBTAG_REGEX = /^([a-z]{2}|[0-9]{3})$/i;
const UNICODE_VARIANT_SUBTAG_REGEX = /^([a-z0-9]{5,8}|[0-9][a-z0-9]{3})$/i;

const TKEY_REGEX = /^[a-z][0-9]$/i;

const SEPARATOR = '-';

function compareKV(t1: Array<any>, t2: Array<any>): number {
  return t1[0] < t2[0] ? -1 : t1[0] > t2[0] ? 1 : 0;
}

function compareExtension(e1: Extension, e2: Extension): number {
  return e1.type < e2.type ? -1 : e1.type > e2.type ? 1 : 0;
}

export function verifyUnicodeLanguageSubtag(lang: string): asserts lang {
  if (!ALPHA_2_3.test(lang) && !ALPHA_5_8.test(lang)) {
    throw new RangeError('Malformed unicode_language_subtag');
  }
}

export function isUnicodeRegionSubtag(region: string): boolean {
  return UNICODE_REGION_SUBTAG_REGEX.test(region);
}

export function isUnicodeScriptSubtag(script: string): boolean {
  return ALPHA_4.test(script);
}

function parseLanguageId(chunks: string[]): UnicodeLanguageId {
  const lang = chunks.shift();
  if (!lang) {
    throw new RangeError('Missing unicode_language_subtag');
  }
  if (lang === 'root') {
    return {lang: 'root', variants: []};
  }
  // unicode_language_subtag
  verifyUnicodeLanguageSubtag(lang);
  let script;
  // unicode_script_subtag
  if (isUnicodeScriptSubtag(chunks[0])) {
    script = chunks.shift();
  }
  let region;
  // unicode_region_subtag
  if (isUnicodeRegionSubtag(chunks[0])) {
    region = chunks.shift();
  }
  const variants: Record<string, any> = {};
  while (chunks.length && UNICODE_VARIANT_SUBTAG_REGEX.test(chunks[0])) {
    const variant = chunks.shift()!;
    if (variant in variants) {
      throw new RangeError(`Duplicate variant "${variant}"`);
    }
    variants[variant] = 1;
  }
  return {
    lang,
    script,
    region,
    variants: Object.keys(variants),
  };
}

function parseUnicodeExtension(chunks: string[]): UnicodeExtension {
  const keywords = [];
  let keyword;
  while (chunks.length && (keyword = parseKeyword(chunks))) {
    keywords.push(keyword);
  }
  if (keywords.length) {
    return {
      type: 'u',
      keywords,
      attributes: [],
    };
  }
  // Mix of attributes & keywords
  // Check for attributes first
  const attributes = [];
  while (chunks.length && ALPHANUM_3_8.test(chunks[0])) {
    attributes.push(chunks.shift()!);
  }
  while (chunks.length && (keyword = parseKeyword(chunks))) {
    keywords.push(keyword);
  }
  if (keywords.length || attributes.length) {
    return {
      type: 'u',
      attributes,
      keywords,
    };
  }
  throw new RangeError('Malformed unicode_extension');
}

function parseKeyword(chunks: string[]): KV | undefined {
  let key;
  if (!KEY_REGEX.test(chunks[0])) {
    return;
  }
  key = chunks.shift()!;

  const type = [];
  while (chunks.length && TYPE_REGEX.test(chunks[0])) {
    type.push(chunks.shift());
  }
  let value: string = '';
  if (type.length) {
    value = type.join(SEPARATOR);
  }
  return [key, value];
}

function parseTransformedExtension(chunks: string[]): TransformedExtension {
  let lang: UnicodeLanguageId | undefined;
  try {
    lang = parseLanguageId(chunks);
  } catch (e) {
    // Try just parsing tfield
  }
  const fields: KV[] = [];
  while (chunks.length && TKEY_REGEX.test(chunks[0])) {
    const key = chunks.shift()!;
    const value = [];
    while (chunks.length && ALPHANUM_3_8.test(chunks[0])) {
      value.push(chunks.shift());
    }
    if (!value.length) {
      throw new RangeError(`Missing tvalue for tkey "${key}"`);
    }
    fields.push([key, value.join(SEPARATOR)]);
  }
  if (fields.length) {
    return {
      type: 't',
      fields,
      lang,
    };
  }
  throw new RangeError('Malformed transformed_extension');
}
function parsePuExtension(chunks: string[]): PuExtension {
  const exts = [];
  while (chunks.length && ALPHANUM_1_8.test(chunks[0])) {
    exts.push(chunks.shift());
  }
  if (exts.length) {
    return {
      type: 'x',
      value: exts.join(SEPARATOR),
    };
  }
  throw new RangeError('Malformed private_use_extension');
}
function parseOtherExtensionValue(chunks: string[]): string {
  const exts = [];
  while (chunks.length && ALPHANUM_2_8.test(chunks[0])) {
    exts.push(chunks.shift());
  }
  if (exts.length) {
    return exts.join(SEPARATOR);
  }
  return '';
}
function parseExtensions(chunks: string[]): Omit<UnicodeLocaleId, 'lang'> {
  if (!chunks.length) {
    return {extensions: []};
  }
  const extensions: UnicodeLocaleId['extensions'] = [];
  let unicodeExtension;
  let transformedExtension;
  let puExtension;
  const otherExtensionMap: Record<string, OtherExtension> = {};
  do {
    const type = chunks.shift()!;
    switch (type) {
      case 'u':
      case 'U':
        if (unicodeExtension) {
          throw new RangeError('There can only be 1 -u- extension');
        }
        unicodeExtension = parseUnicodeExtension(chunks);
        extensions.push(unicodeExtension);
        break;
      case 't':
      case 'T':
        if (transformedExtension) {
          throw new RangeError('There can only be 1 -t- extension');
        }
        transformedExtension = parseTransformedExtension(chunks);
        extensions.push(transformedExtension);
        break;
      case 'x':
      case 'X':
        if (puExtension) {
          throw new RangeError('There can only be 1 -x- extension');
        }
        puExtension = parsePuExtension(chunks);
        extensions.push(puExtension);
        break;
      default:
        if (!OTHER_EXTENSION_TYPE.test(type)) {
          throw new RangeError('Malformed extension type');
        }
        if (type in otherExtensionMap) {
          throw new RangeError(`There can only be 1 -${type}- extension`);
        }
        const extension: OtherExtension = {
          type: type as 'a',
          value: parseOtherExtensionValue(chunks),
        };
        otherExtensionMap[extension.type] = extension;
        extensions.push(extension);
        break;
    }
  } while (chunks.length);
  return {extensions};
}

function canonicalizeAttrs(strs: string[]): string[] {
  return Object.keys(
    strs.reduce((all: Record<string, number>, str) => {
      all[str.toLowerCase()] = 1;
      return all;
    }, {})
  ).sort();
}

function canonicalizeKVs(arr: KV[]): KV[] {
  const all: Record<string, any> = {};
  const result: KV[] = [];
  for (const kv of arr) {
    if (kv[0] in all) {
      continue;
    }
    all[kv[0]] = 1;
    if (!kv[1] || kv[1] === 'true') {
      result.push([kv[0].toLowerCase()]);
    } else {
      result.push([kv[0].toLowerCase(), kv[1].toLowerCase()]);
    }
  }
  return result.sort(compareKV);
}

export function parse(locale: string): UnicodeLocaleId {
  const chunks = locale.split(SEPARATOR);
  const lang = parseLanguageId(chunks);
  return canonicalize({
    lang,
    ...parseExtensions(chunks),
  });
}

export function canonicalizeUnicodeLanguageId(lang: UnicodeLanguageId): void {
  lang.lang = lang.lang.toLowerCase();
  if (lang.script) {
    lang.script =
      lang.script[0].toUpperCase() + lang.script.slice(1).toLowerCase();
  }
  if (lang.region) {
    lang.region = lang.region.toUpperCase();
  }
  if (lang.variants) {
    lang.variants = lang.variants.map(v => v.toLowerCase()).sort();
  }
}

/**
 * Canonicalize based on
 * https://www.unicode.org/reports/tr35/tr35.html#Canonical_Unicode_Locale_Identifiers
 * https://tc39.es/ecma402/#sec-canonicalizeunicodelocaleid
 * IMPORTANT: This modifies the object inline
 * @param locale
 */
export function canonicalize(locale: UnicodeLocaleId): UnicodeLocaleId {
  canonicalizeUnicodeLanguageId(locale.lang);
  if (locale.extensions) {
    for (const extension of locale.extensions) {
      switch (extension.type) {
        case 'u':
          extension.keywords = canonicalizeKVs(extension.keywords);
          if (extension.attributes) {
            extension.attributes = canonicalizeAttrs(extension.attributes);
          }
          break;
        case 't':
          if (extension.lang) {
            canonicalizeUnicodeLanguageId(extension.lang);
          }
          extension.fields = canonicalizeKVs(extension.fields);
          break;
        default:
          extension.value = extension.value.toLowerCase();
          break;
      }
    }
    locale.extensions.sort(compareExtension);
  }

  return locale;
}
