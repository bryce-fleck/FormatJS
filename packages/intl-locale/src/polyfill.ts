import {IntlLocale} from './';
if (!('Locale' in Intl)) {
  Object.defineProperty(Intl, 'Locale', {
    value: IntlLocale,
    writable: true,
    enumerable: false,
    configurable: true,
  });
}
