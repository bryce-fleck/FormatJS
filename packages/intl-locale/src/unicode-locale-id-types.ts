export interface UnicodeLocaleId {
  lang: UnicodeLanguageId;
  unicodeExtension?: UnicodeExtension;
  transformedExtension?: TransformedExtension;
  puExtension?: PuExtension;
  otherExtensions?: OtherExtensions;
}

export interface UnicodeLanguageId {
  lang: string;
  script?: string;
  region?: string;
  variants?: string[];
}

export type KV = [string, string];

export interface UnicodeExtension {
  type: 'u';
  keywords: KV[];
  attributes?: string[];
}

export interface TransformedExtension {
  type: 't';
  fields: KV[];
  lang?: UnicodeLanguageId;
}
export interface PuExtension {
  type: 'x';
  value: string;
}

export type OtherExtensions = Record<string, string>;
