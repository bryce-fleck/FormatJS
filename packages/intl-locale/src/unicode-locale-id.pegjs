{
    function ignoreSep ([sep, str]) {
        return str
    }

    function titleCase(str) {
        return str[0].toUpperCase() + str.slice(1).toLowerCase()
    }

    function canonicalizeFields(fields) {
        return fields.map(ignoreSep).filter(f => f !== 'true').sort(compareTuple)
    }

    function compareTuple(t1, t2) {
        return t1[0] < t2[0] ? -1 : t1[0] > t2[0] ? 1 : 0
    }

    function canonicalizeLangId (languages, region, variants) {
        const [lang, script] = languages
      const result = {
          lang,
          script,
          region: '',
          variants: []
      }
      if (lang.length === 4) {
          result.script = titleCase(lang)
      } else {
          result.lang = lang.toLowerCase()
          if (script) {
              result.script = titleCase(script[1])
          }
      }

      if (region) {
          result.region = region[1].toUpperCase()
      }

      if (Array.isArray(variants) && variants.length) {
          result.variants = variants.map(ignoreSep).sort()
      }
      return result
    }
}

// https://unicode.org/reports/tr35/#unicode_locale_id
// This does not contain unicode_subdivision_id, unicode_subdivision_suffix & unicode_measure_unit
unicode_locale_id = lang:unicode_language_id extensions:extensions* puExtension:pu_extensions? {
    let unicodeExtension
    let transformedExtension
    const otherExtensions = {}
    for (const ext of extensions) {
        if (ext.type === 'u') {
            if (unicodeExtension) {
                throw new RangeError('There can only be 1 -u- extension')
            }
            unicodeExtension = ext
        } else if (ext.type === 't') {
            if (transformedExtension) {
                throw new RangeError('There can only be 1 -t- extension')
            }
            transformedExtension = ext
        } else {
            if (otherExtensions[ext.type]) {
                throw new RangeError(`There can only be 1 -${ext.type}- extension`)   
            }
            otherExtensions[ext.type] = ext.value
        }
    }
    return {
        lang,
        otherExtensions,
        unicodeExtension,
        transformedExtension,
        puExtension
    }
};
extensions = unicode_locale_extensions
    / transformed_extensions
    / other_extensions;
unicode_locale_extensions = sep [uU]
    extension:(
        attributes:(sep attribute)+ keywords:(sep keyword)* {
            return {
                attributes: attributes.map(ignoreSep).sort(),
                keywords: Array.isArray(keywords) ? keywords.map(ignoreSep).sort(compareTuple) : []
            }
        }
        / keywords:(sep keyword)+ {
            return { keywords: keywords.map(ignoreSep).sort(compareTuple) }
        }
    ) {
        return {...extension, type: 'u'}
    };
transformed_extensions = sep [tT]
   extension:(
       sep lang:tlang fields:(sep tfield)* { 
           const result = {lang, fields: []}
           if (Array.isArray(fields)) {
               result.fields = canonicalizeFields(fields)
           }
       }
        / fields:(sep tfield)+ {
            return {fields: canonicalizeFields(fields) }
        }
    ) {
        return {...extension, type: 't'}
    };
pu_extensions = sep [xX] extensions:(sep anum_1_8)+ {
    return { type: 'x', values: extensions.map(ignoreSep).join('-') }
};
// alphanum-[tTuUxX] but there's no exclusion in pegjs
other_extensions = sep type:[0-9a-sA-SvwyzVWYZ] values:(sep anum_2_8)+ {
    return { 
        type,
        value: values.map(ignoreSep).join('-') 
    }
};
keyword = key:key value:(sep type)? {
    if (Array.isArray(value) && value[1]) {
        return [key, value[1] === 'true' ? '' : value[1]]
    }
    return [key, '']
};
key = $(anum alpha);

type = $(anum_3_8 (sep anum_3_8)*);	
attribute = anum_3_8;
tlang = lang:unicode_language_subtag
  script:(sep unicode_script_subtag)?
  region:(sep unicode_region_subtag)?
  variants:(sep unicode_variant_subtag)* {
      return canonicalizeLangId([lang, script], region, variants)
  };
tfield = k:tkey v:tvalue {
    return [k, v]
};	
tkey = $(alpha digit);
tvalue = val:(sep anum_3_8)+ {
    return val.map(ignoreSep)
};

unicode_language_id	= "root" { return { lang: 'root' }} 
  / languages:(unicode_language_subtag (sep unicode_script_subtag)? / unicode_script_subtag) region:(sep unicode_region_subtag)? variants:(sep unicode_variant_subtag)* {
      return canonicalizeLangId(languages, region, variants)
  };
unicode_language_subtag	= alpha_5_8 / alpha_2_3

unicode_script_subtag = $(alpha alpha alpha alpha);	
unicode_region_subtag = $((alpha alpha) / (digit digit digit));
unicode_variant_subtag = $( 
      anum_5_8
    / digit anum anum anum);

// The section below are repetitive bc pegjs doesn't support range (e.g alpha{3,8})
alpha_2_3 = $(
    alpha alpha alpha
    / alpha alpha 
)

alpha_5_8 = $(
    alpha alpha alpha alpha alpha alpha alpha alpha 
    / alpha alpha alpha alpha alpha alpha alpha
    / alpha alpha alpha alpha alpha alpha
    / alpha alpha alpha alpha alpha
)

anum_1_8 = 
    $(anum_2_8 / anum)

anum_2_8 = 
    $(anum_3_8 / anum anum)

anum_1_4 = 
    $(anum anum anum anum 
    / anum anum anum
    / anum anum
    / anum)

anum_3_8 = 
    $(anum_5_8
    / anum anum anum anum
    / anum anum anum)
    
anum_5_8 = 
    $(anum anum anum anum anum anum anum anum
    / anum anum anum anum anum anum anum
    / anum anum anum anum anum anum
    / anum anum anum anum anum)

// Basic char defs
sep	= "-";
digit = [0-9];
alpha = [A-Za-z];
anum = [0-9A-Za-z];
