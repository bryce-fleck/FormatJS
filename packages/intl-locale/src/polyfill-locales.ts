import './polyfill';
import {Locale} from '.';
import * as data from 'cldr-core/supplemental/likelySubtags.json';
const PolyfilledLocale = (Intl as any).Locale as typeof Locale;
if (typeof PolyfilledLocale._addLikelySubtagData === 'function') {
  PolyfilledLocale._addLikelySubtagData(data.supplemental.likelySubtags);
}
