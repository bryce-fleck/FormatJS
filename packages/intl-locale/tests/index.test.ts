import {Locale} from '../src';
import {parse} from '../src/parser';

describe('intl-locale', () => {
  it('ast', function() {
    expect(parse('en-u-foo-bar-nu-thai-ca-buddhist-kk-true')).toMatchSnapshot();
    expect(parse('da-u-ca-gregory-ca-buddhist')).toMatchSnapshot();
  });
  it('toString', function() {
    // expect(() => IntlLocale.prototype.toString.call(IntlLocale.prototype)).toThrowError(TypeError)
    expect(
      new Locale('en-u-foo-bar-nu-thai-ca-buddhist-kk-true').toString()
    ).toBe('en-u-bar-foo-ca-buddhist-kk-nu-thai');
  });
  it.skip('canonicalizes twice', function() {
    expect(new Locale('und-Armn-SU', {language: 'ru'}).toString()).toBe(
      'ru-Armn-AM'
    );
  });
});
