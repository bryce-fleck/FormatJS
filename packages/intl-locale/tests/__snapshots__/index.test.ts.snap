// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intl-locale ast 1`] = `
Object {
  "extensions": Array [
    Object {
      "attributes": Array [
        "bar",
        "foo",
      ],
      "keywords": Array [
        Array [
          "ca",
          "buddhist",
        ],
        Array [
          "kk",
          "",
        ],
        Array [
          "nu",
          "thai",
        ],
      ],
      "type": "u",
    },
  ],
  "lang": Object {
    "lang": "en",
    "region": undefined,
    "script": undefined,
    "variants": Array [],
  },
}
`;

exports[`intl-locale ast 2`] = `
Object {
  "extensions": Array [
    Object {
      "attributes": Array [],
      "keywords": Array [
        Array [
          "ca",
          "gregory",
        ],
      ],
      "type": "u",
    },
  ],
  "lang": Object {
    "lang": "da",
    "region": undefined,
    "script": undefined,
    "variants": Array [],
  },
}
`;
